mmcv/ops/cc_attention.py

```diff
@@ -1,91 +1,79 @@
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-from torch.autograd.function import once_differentiable
+from einops import rearrange
 
-from mmcv.cnn import Scale
-from ..utils import ext_loader
 
-ext_module = ext_loader.load_ext(
-    '_ext', ['ca_forward', 'ca_backward', 'ca_map_forward', 'ca_map_backward'])
+def NEG_INF_DIAG(n, device):
+    """Returns a diagonal matrix of size [n, n].
 
-
-class CAWeightFunction(torch.autograd.Function):
-
-    @staticmethod
-    def symbolic(g, t, f):
-        return g.op('MMCVCAWeight', t, f)
-
-    @staticmethod
-    def forward(ctx, t, f):
-        n, c, h, w = t.size()
-        weight = torch.zeros(n, h + w - 1, h, w).to(t.device)
-        ext_module.ca_forward(t, f, weight)
-
-        ctx.save_for_backward(t, f)
-
-        return weight
-
-    @staticmethod
-    @once_differentiable
-    def backward(ctx, dw):
-        t, f = ctx.saved_tensors
-        dt = torch.zeros_like(t)
-        df = torch.zeros_like(f)
-        ext_module.ca_backward(dw, t, f, dt, df)
-        return dt, df
-
-
-class CAMapFunction(torch.autograd.Function):
-
-    @staticmethod
-    def symbolic(g, weight, v):
-        return g.op('MMCVCAMap', weight, v)
-
-    @staticmethod
-    def forward(ctx, weight, v):
-        out = torch.zeros_like(v)
-        ext_module.ca_map_forward(weight, v, out)
-
-        ctx.save_for_backward(weight, v)
-
-        return out
-
-    @staticmethod
-    @once_differentiable
-    def backward(ctx, dout):
-        weight, v = ctx.saved_tensors
-        dw = torch.zeros_like(weight)
-        dv = torch.zeros_like(v)
-        ext_module.ca_map_backward(dout, weight, v, dw, dv)
-
-        return dw, dv
-
-
-ca_weight = CAWeightFunction.apply
-ca_map = CAMapFunction.apply
+    The diagonal are all "-inf". This is for avoiding calculating the
+    overlapped element in the Criss-Cross twice.
+    """
+    return torch.diag(torch.tensor(float('-inf')).to(device).repeat(n), 0)
 
 
 class CrissCrossAttention(nn.Module):
-    """Criss-Cross Attention Module."""
+    """Criss-Cross Attention Module.
+
```

Comment:
Suggest to also update arguments in the docstring.

Reply:
OK.